### PR TITLE
WIP: fix query exception when attendee.birthdate is empty

### LINKED ISSUE #1882: Session::guess_attendee_watchentry() causes c.BADGE_LOCK error message
I find the most interesting things when I have the debugger open... :)

This one is ultimately harmless but would be cool to fix if it's easy.  I have a fix.

backstory:

---

We have a custom django tag called 'nav_menu'.  In certain registration forms, we include this tag similarly to this in registration/menu.html:

```
{% nav "watchlist?attendee_id={id}" "Watch List Entry" attendee.banned %}
```

This calls 'attendee.banned' to determine if it should display this attendee's watchlist record, which calls Attendee.watchlist_guest() which looks like this:

``` python
    @property
    def watchlist_guess(self):
        try:
            with Session() as session:
                return [w.to_dict() for w in session.guess_attendee_watchentry(self)]
        except:
            return None
```

I looked into why this needed to be locked into such a huge exception block, and discovered that session.guess_attendee_watchentry() seems to be able to generate a query that fails if the attendee doesn't have a birthdate set.  Presumably, this exception wrap was added to catch that case.

So, this all works but has two side effects:
1) By swallowing up the exception, the original error message from the exception (which pointed out that the birthdate format being generated in the query was invalid) was omitted, making it hard to fix the underlying issue.

2) On my box (and I hadn't noticed this before but...) I receive this error:

```
2016-06-21 15:59:40,509 [WARNING] uber.models: failed releasing c.BADGE_LOCK on db error; these errors should not happen in the first place and we do not expect releasing the lock to fail when they do, but we still want to keep going if/when this does occur
```

This is however, harmless, just spammy.  It's because this function is registered to be called anytime there's a sqlalchemy error:

```
def _release_badge_lock_on_error(*args, **kwargs):
    try:
        c.BADGE_LOCK.release()
    except:
        log.warn('failed releasing c.BADGE_LOCK on db error; these errors should not happen in the first place and we do not expect releasing the lock to fail when they do, but we still want to keep going if/when this does occur')
```

Which just REALLY wants to make sure that c.BADGE_LOCK is released, even if it was never locked.  Which, is totally sane.

But, if we can prevent the exception in the first place, we won't hit this side effect error message.

### PR DESCRIPTION
**Pull request by binary1230**
This query could fail if the attendee in question didn't have a birthdate set to non-empty/non-null value.  We now check if they have a valid birthdate before including the part of the query that checks the birthdate.

As a result of fixing this I think it's probably the right move to remove the aggressive try/except block in the calling function and fix any errors that come up if/when they come up.

fixes #1882. see that issue for a detailed description of a bunch of non-obvious stuff.

Needs some eyes on it and probably more testing.

### CODE REVIEW

File: uber/models.py [hunk 1]
```diff
@@ -473,11 +473,20 @@ def jobs_for_signups(self):
...
-                                                     WatchList.active == True)).all()
+            return self.query(WatchList).filter(
+                    and_(
+                        or_(
+                            WatchList.first_names.contains(attendee.first_name),
+                            and_(WatchList.email != '', WatchList.email == attendee.email),
+                            *[WatchList.birthdate == attendee.birthdate] if
+                                WatchList.birthdate and WatchList.birthdate != '' and
+                                attendee.birthdate and attendee.birthdate != ''
+                            else []
```
Comment (binary1230) [inline uber/models.py]: this syntax is a bit weird to look at at first, but what it's doing is like the following:

ok_to_check_for_birthdate = WatchList.birthdate and WatchList.birthdate != '' and attendee.birthdate and attendee.birthdate != ''

if ok_to_check_for_birthdate:
   include this check in SQL: WatchList.birthdate == attendee.birthdate
if not, don't include it.

**Comment by kitsuta** [comment]
Looks good. +1

File: uber/models.py [hunk 2]
```diff
@@ -473,11 +473,20 @@ def jobs_for_signups(self):
...
-                                                         and_(WatchList.birthdate != None, WatchList.birthdate == attendee.birthdate)),
-                                                     WatchList.last_name == attendee.last_name,
-                                                     WatchList.active == True)).all()
+            return self.query(WatchList).filter(
+                    and_(
+                        or_(
+                            WatchList.first_names.contains(attendee.first_name),
+                            and_(WatchList.email != '', WatchList.email == attendee.email),
+                            *[WatchList.birthdate == attendee.birthdate] if
+                                WatchList.birthdate and WatchList.birthdate != '' and
```
Comment (EliAndrewC) [inline uber/models.py]: This line is unnecessary:

```python
>>> WatchList.birthdate
<sqlalchemy.orm.attributes.InstrumentedAttribute object at 0x7f364dc3ec50>
```

Remember that ``attendee`` is a specific instance of the ``Attendee`` class whose birthdate can have a particular value, but ``WatchList.birthdate`` is the instrumented attribute for the whole class and will never have a particular value such as ``None`` or ``''``.

Comment (kitsuta) [inline uber/models.py]: Wait, aren't we essentially iterating through rows in a query, here? So we would want to keep `!= ''` to throw out WatchList entries with blank data?

Comment (EliAndrewC) [inline uber/models.py]: If that's what we mean to do then we need to move that condition out of the ``if`` block and into the ``*[]`` block.

**Comment by RobRuana** [comment]
Whoops! I tried to rebase Dom's work on top of our current master `git rebase master` But then when I did a `git pull` it merged in the whole history of master. Instead of trying to fix it, I am going to close this pull request and open a new one.

**Pull request closed without merging**